relays/ethereum/src/rpc.rs

```diff
@@ -0,0 +1,338 @@
+		fn get_transaction_receipt(transaction_hash: Params) -> Receipt;
+		#[rpc(method = "eth_getTransactionCount")]
+		fn get_transaction_count(address: Params) -> U256;
+		#[rpc(method = "eth_submitTransaction")]
+		fn submit_transaction(transaction: Params) -> EthereumTxHash;
+		#[rpc(method = "eth_call")]
+		fn call(transaction_call: Params) -> Bytes;
+	}
+
+	Substrate {
+		#[rpc(method = "chain_getHeader")]
+		fn chain_get_header(params: Params) -> SubstrateHeader;
+		#[rpc(method = "chain_getBlockHash")]
+		fn chain_get_block_hash(block_number: Params) -> SubstrateHash;
+		#[rpc(method = "system_accountNextIndex")]
+		fn system_account_next_index(account_id: Params) -> node_primitives::Index;
+		#[rpc(method = "author_submitExtrinsic")]
+		fn author_submit_extrinsic(extrinsic: Params) -> SubstrateHash;
+		#[rpc(method = "state_call")]
+		fn state_call(method: Params) -> Bytes;
+	}
+}
+
+/// The API for the supported Ethereum RPC methods.
+#[async_trait]
+pub trait EthereumRpc {
+	async fn estimate_gas(&mut self, call_request: CallRequest) -> Result<U256>;
+	async fn best_block_number(&mut self) -> Result<u64>;
+	async fn header_by_number(&mut self, block_number: u64) -> Result<EthereumHeader>;
+	async fn header_by_hash(&mut self, hash: H256) -> Result<EthereumHeader>;
+	async fn transaction_receipt(&mut self, transaction_hash: H256) -> Result<Receipt>;
+	async fn account_nonce(&mut self, address: EthAddress) -> Result<U256>;
+	async fn submit_transaction(&mut self, signed_raw_tx: SignedRawTx) -> Result<EthereumTxHash>;
+	async fn eth_call(&mut self, call_transaction: CallRequest) -> Result<Bytes>;
+}
+
+/// The client used to interact with an Ethereum node through RPC.
+pub struct EthereumRpcClient {
+	client: RawClient<HttpTransportClient>,
+}
+
+impl EthereumRpcClient {
+	/// Create a new Ethereum RPC Client.
+	pub fn new(params: EthereumConnectionParams) -> Self {
+		let uri = format!("http://{}:{}", params.host, params.port);
+		let transport = HttpTransportClient::new(&uri);
+		let client = RawClient::new(transport);
+
+		Self { client }
+	}
+}
+
+#[async_trait]
+impl EthereumRpc for EthereumRpcClient {
+	/// Estimate gas usage for the given call.
+	async fn estimate_gas(&mut self, call_request: CallRequest) -> Result<U256> {
+		let params = Params::Array(vec![serde_json::to_value(call_request)?]);
+		Ok(Ethereum::estimate_gas(&mut self.client, params).await?)
+	}
+
+	/// Retrieve number of the best known block from the Ethereum node.
+	async fn best_block_number(&mut self) -> Result<u64> {
+		Ok(Ethereum::block_number(&mut self.client).await?.as_u64())
+	}
+
+	/// Retrieve block header by its number from Ethereum node.
+	async fn header_by_number(&mut self, block_number: u64) -> Result<EthereumHeader> {
+		// Only want to get hashes back from the RPC
+		let return_full_tx_obj = false;
+
+		let params = Params::Array(vec![
+			serde_json::to_value(U64::from(block_number)).expect(INT_SERIALIZATION_PROOF),
```

Comment:
Why not simply use `?` here instead of `expect`? I feel the code would simpler and I see no advantages of using `expect`.